From 6b34c86836ba3b5135aadb4b6aa00ab4c65943fb Mon Sep 17 00:00:00 2001
From: Christian Marangi <ansuelsmth@gmail.com>
Date: Mon, 3 Jul 2023 15:55:43 +0200
Subject: [PATCH 11/20] net: dsa: qca8k: restore default VID with vlan 0

Qca8k sets a default VID for untagged frame to handle them internally.
This default VID is currently set in qca8k_setup but is overwrite when a
vlan for the port is set and never restored when the port is configured
again for untagged frames (vlan 0).

Rework the vlan_add function to handle this case and always set the
default VID and assign the default VID if we are in vlan 0 case.

Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>
---
 drivers/net/dsa/qca/qca8k-common.c | 24 +++++++++++++-----------
 1 file changed, 13 insertions(+), 11 deletions(-)

--- a/drivers/net/dsa/qca/qca8k-common.c
+++ b/drivers/net/dsa/qca/qca8k-common.c
@@ -1021,25 +1021,27 @@ int qca8k_port_vlan_add(struct dsa_switc
 	bool untagged = vlan->flags & BRIDGE_VLAN_INFO_UNTAGGED;
 	bool pvid = vlan->flags & BRIDGE_VLAN_INFO_PVID;
 	struct qca8k_priv *priv = ds->priv;
+	u16 vid = vlan->vid;
 	int ret;
 
-	ret = qca8k_vlan_add(priv, port, vlan->vid, untagged);
+	ret = qca8k_vlan_add(priv, port, vid, untagged);
 	if (ret) {
 		dev_err(priv->dev, "Failed to add VLAN to port %d (%d)", port, ret);
 		return ret;
 	}
 
-	if (pvid) {
-		ret = qca8k_rmw(priv, QCA8K_EGRESS_VLAN(port),
-				QCA8K_EGREES_VLAN_PORT_MASK(port),
-				QCA8K_EGREES_VLAN_PORT(port, vlan->vid));
-		if (ret)
-			return ret;
+	if (!pvid)
+		vid = QCA8K_PORT_VID_DEF;
 
-		ret = qca8k_write(priv, QCA8K_REG_PORT_VLAN_CTRL0(port),
-				  QCA8K_PORT_VLAN_CVID(vlan->vid) |
-				  QCA8K_PORT_VLAN_SVID(vlan->vid));
-	}
+	ret = qca8k_rmw(priv, QCA8K_EGRESS_VLAN(port),
+			QCA8K_EGREES_VLAN_PORT_MASK(port),
+			QCA8K_EGREES_VLAN_PORT(port, vid));
+	if (ret)
+		return ret;
+
+	ret = qca8k_write(priv, QCA8K_REG_PORT_VLAN_CTRL0(port),
+			  QCA8K_PORT_VLAN_CVID(vid) |
+			  QCA8K_PORT_VLAN_SVID(vid));
 
 	return ret;
 }
